nix/redis-cluster.nix

```diff
@@ -0,0 +1,138 @@
+
+        If master node is down, a replica node will take over.
+      '';
+    };
+
+    replicas = lib.mkOption {
+      type = types.int;
+      default = 1;
+      description = ''
+        Number of replicas per Master node.
+      '';
+    };
+
+    outputs.settings = lib.mkOption {
+      type = types.deferredModule;
+      internal = true;
+      readOnly = true;
+      default =
+        let
+          mkNodeProcess = nodeConfig:
+            let
+              port = builtins.toString nodeConfig.port;
+              redisConfig = pkgs.writeText "redis.conf" ''
+                port ${port}
+                cluster-enabled yes
+                cluster-config-file nodes-${port}.conf
+                cluster-node-timeout ${builtins.toString config.timeout}
+                appendonly yes
+                appendfilename "appendonly-${port}.aof"
+                dbfilename "dump-${port}.rdb"
+
+                ${lib.optionalString (config.bind != null) "bind ${config.bind}"}
+                ${nodeConfig.extraConfig}
+              '';
+
+              startScript = pkgs.writeShellScriptBin "start-redis" ''
+                set -euo pipefail
+
+                export REDISDATA=${config.dataDir}
+
+                if [[ ! -d "$REDISDATA" ]]; then
+                  mkdir -p "$REDISDATA"
+                fi
+
+                exec ${config.package}/bin/redis-server ${redisConfig} --dir "$REDISDATA"
+              '';
+            in
+            {
+              "${name}-node-${port}" = {
+                command = "${startScript}/bin/start-redis";
+                shutdown.command = "${config.package}/bin/redis-cli -p ${port} shutdown nosave";
+
+                readiness_probe = {
+                  exec.command = "${config.package}/bin/redis-cli -p ${port} ping";
+                  initial_delay_seconds = 2;
+                  period_seconds = 10;
+                  timeout_seconds = 4;
+                  success_threshold = 1;
+                  failure_threshold = 5;
+                };
+
+                # https://github.com/F1bonacc1/process-compose#-auto-restart-if-not-healthy
+                availability.restart = "on_failure";
+              };
+            };
+          hosts = builtins.map (node: "${config.bind}:${builtins.toString node.port}") config.nodes;
+          healthyNodes = builtins.map (node: { "${name}-node-${builtins.toString node.port}".condition = "process_healthy"; }) config.nodes;
+          createClusterProcess = {
+            "${name}-cluster-create" = {
+              depends_on = lib.mkMerge healthyNodes;
+              command = "${config.package}/bin/redis-cli --cluster create ${lib.concatStringsSep " " hosts} --cluster-replicas ${builtins.toString config.replicas} --cluster-yes";
+            };
+          };
+          processesList = (builtins.map mkNodeProcess config.nodes) ++ [ createClusterProcess ];
+        in
+        {
+          processes = lib.mkMerge processesList;
```

Comment:
How does this behave when there is a key conflict (by accident/bug)? How does `mkMerge` compare to using `imports`?

Reply:
Why not inline it simply?

```nix
{
  processes = lib.mapAttrs mkNodeProcess (listToAttrs (_ config.nodes)) // {
    "${name}-cluster-create" = clusterProcess;
  };
}
```

Reply:
> How does this behave when there is a key conflict (by accident/bug)?

Case 1: 
```
{
  config.nodes = [
    { port = 30001; }
    { port = 30002; }
    { port = 30001; }
  ];
}
```
Case 2: 
```
{
  config.nodes = [
    { port = 30001; }
    { port = 30002; }
    { port = 30001; extraConfig = ''maxmemory 2mb''; }
  ];
}
```
In `Case 1` `mkMerge` is considering the first element of the `config.nodes` to resolve the conflict whereas in `Case 2` it will error out saying that "The option `perSystem.aarch64-darwin.process-compose.default.settings.processes.c1-node-30002.command' has conflicting definition values".

Reply:
> Why not inline it simply?
```nix
{
  processes = lib.mapAttrs mkNodeProcess (listToAttrs (_ config.nodes)) // {
    "${name}-cluster-create" = clusterProcess;
  };
}
```

This can be done, but the problem will be that in the `Case 2` mentioned above it will never throw an error while trying to `nix run` and node `{ port = 30001; extraConfig = ''maxmemory 2mb''; }` will override `{ port = 30001; }`

Reply:
Decided to go with `attrsOf submodule` instead of `listOf submodule` for `nodes`, as suggested by @srid